### PR TITLE
3176 improve new parameter optimization classes for symbolic regression

### PR DESCRIPTION
**Pull request by mkommend**

### CODE REVIEW

File: HeuristicLab.Problems.DataAnalysis.Symbolic/3.4/Interpreter/NativeInterpreter.cs [hunk 1]
```diff
@@ -80,21 +80,20 @@ public sealed class NativeInterpreter : ParameterizedNamedItem, ISymbolicDataAna
...
           code[i].Coeff = number.Value;
-        } else if (n is ConstantTreeNode constant) {
+        } else if (nodes[i] is ConstantTreeNode constant) {
           code[i].Coeff = constant.Value;
         }
-        --i;
       }
       // second pass to calculate lengths
-      for (i = 0; i < code.Length; i++) {
+      for (int i = 0; i < code.Length; i++) {
```
Comment (foolnotion) [inline HeuristicLab.Problems.DataAnalysis.Symbolic/3.4/Interpreter/NativeInterpreter.cs:96]: the second pass is no longer needed

File: HeuristicLab.Problems.DataAnalysis.Symbolic/3.4/Interpreter/NativeInterpreter.cs [hunk 2]
```diff
@@ -80,21 +80,20 @@ public sealed class NativeInterpreter : ParameterizedNamedItem, ISymbolicDataAna
       nodes = root.IterateNodesPrefix().ToList(); nodes.Reverse();
       var code = new NativeInstruction[nodes.Count];
       if (root.SubtreeCount > ushort.MaxValue) throw new ArgumentException("Number of subtrees is too big (>65.535)");
-      int i = code.Length - 1;
-      foreach (var n in root.IterateNodesPrefix()) {
-        code[i] = new NativeInstruction { Arity = (ushort)n.SubtreeCount, OpCode = opCodeMapper(n), Length = 1, Optimize = 0 };
-        if (n is VariableTreeNode variable) {
+
+      for (int i = 0; i < code.Length; i++) {
+        code[i] = new NativeInstruction { Arity = (ushort)nodes[i].SubtreeCount, OpCode = opCodeMapper(nodes[i]), Length = 1, Optimize = 0 };
```
Comment (foolnotion) [inline HeuristicLab.Problems.DataAnalysis.Symbolic/3.4/Interpreter/NativeInterpreter.cs:85]: we can just use `Length = nodes[i].GetLength()` and avoid a second pass below

**Review by foolnotion — CHANGES_REQUESTED** [review]
Looks good! Only one minor thing: I think we can remove the second pass updating lengths in `NativeInterpreter.cs` line 97 and instead use:

```csharp
for (int i = 0; i < code.Length; i++) {
  code[i] = new NativeInstruction { Arity = (ushort)nodes[i].SubtreeCount, OpCode = opCodeMapper(nodes[i]), Length = nodes[i].GetLength(), Optimize = 0 };
  if (nodes[i] is VariableTreeNode variable) {
    code[i].Coeff = variable.Weight;
    code[i].Data = cachedData[variable.VariableName].AddrOfPinnedObject();
  } else if (nodes[i] is NumberTreeNode number) {
    code[i].Coeff = number.Value;
  } else if (nodes[i] is ConstantTreeNode constant) {
    code[i].Coeff = constant.Value;
  }
}
```

File: HeuristicLab.Problems.DataAnalysis.Symbolic/3.4/Interpreter/NativeInterpreter.cs (hunk 2 above)
Comment (mkommend) [reply HeuristicLab.Problems.DataAnalysis.Symbolic/3.4/Interpreter/NativeInterpreter.cs:85]: The suggested change is not possible as it results in evaluation errors. I have tested the changes and afterwards the unit test `TestInterpretersEstimatedValuesConsistency `fails for tree evaluations that would result in infinity, e.g. trees such as LOG(EXP(6.14537408487462 * 'x23' - EXP(7.45401367061198 * 'x12'))).

I haven't found the root cause for that, but it is definitely related to the second pass adjusting the code lengths.

**Comment by mkommend** [comment]
Furthermore we should adapt the provide sample so that the new paremeter optimization evaluation is used and its functionality is covered by an unit test.

File: HeuristicLab.Problems.DataAnalysis.Symbolic/3.4/Interpreter/NativeInterpreter.cs (hunk 2 above)
Comment (foolnotion) [reply HeuristicLab.Problems.DataAnalysis.Symbolic/3.4/Interpreter/NativeInterpreter.cs:85]: I believe this is actually an unrelated issue which occurred by coincidence while you were testing this change. The problem can be reproduced with the original code as well, if you repeat the unit test a few times. As you already noted, this seems to be an inconsistency between how C# and C++ handle extreme values. For example the expression:
`log(exp(5.4×exp(exp(5.6014)) / −4.4952))` evaluates to `−4.885 162 014 × 10^117` in C++ and to `-Infinity` in C#.

**Pull request merged**